rust/src/message/message.rs

```diff
@@ -0,0 +1,178 @@
+    mask::{Mask, MaskedModel},
+    message::{
+        DecodeError,
+        FromBytes,
+        Header,
+        HeaderOwned,
+        Payload,
+        PayloadOwned,
+        Sum2Owned,
+        SumOwned,
+        Tag,
+        ToBytes,
+        UpdateOwned,
+    },
+    LocalSeedDict,
+};
+
+/// A message
+pub struct Message<C, D, M, N> {
+    /// Message header
+    pub header: Header<C>,
+    /// Message payload
+    pub payload: Payload<D, M, N>,
+}
+
+pub type MessageOwned = Message<Certificate, LocalSeedDict, MaskedModel, Mask>;
+
+macro_rules! impl_new {
+	($name:ident, $payload:ty, $tag:expr) => {
+        paste::item! {
+            pub fn [<new_ $name>](coordinator_pk: $crate::CoordinatorPublicKey, participant_pk: $crate::ParticipantPublicKey, payload: $payload) -> Self {
+                Self {
+                    header: Header {
+                        coordinator_pk,
+                        participant_pk,
+                        tag: $tag,
+                        certificate: None,
+                    },
+                    payload: $crate::message::Payload::from(payload),
+                }
+            }
+        }
+    };
+}
+
+impl<C, D, M, N> Message<C, D, M, N>
+where
+    C: Borrow<Certificate>,
+    D: Borrow<LocalSeedDict>,
+    M: Borrow<MaskedModel>,
+    N: Borrow<Mask>,
+{
+    impl_new!(sum, crate::message::Sum, Tag::Sum);
+    impl_new!(update, crate::message::Update<D, M>, Tag::Update);
+    impl_new!(sum2, crate::message::Sum2<N>, Tag::Sum2);
+}
+
+impl<C, D, M, N> ToBytes for Message<C, D, M, N>
+where
+    C: Borrow<Certificate>,
+    D: Borrow<LocalSeedDict>,
+    M: Borrow<MaskedModel>,
+    N: Borrow<Mask>,
+{
+    fn buffer_length(&self) -> usize {
+        self.header.buffer_length() + self.payload.buffer_length()
+    }
+
+    fn to_bytes<T: AsMut<[u8]>>(&self, buffer: &mut T) {
+        self.header.to_bytes(buffer);
+        let mut payload_slice = &mut buffer.as_mut()[self.header.buffer_length()..];
+        self.payload.to_bytes(&mut payload_slice);
+    }
+}
+
+impl FromBytes for MessageOwned {
+    fn from_bytes<T: AsRef<[u8]>>(buffer: &T) -> Result<Self, DecodeError> {
+        let header = HeaderOwned::from_bytes(&buffer)?;
+        let payload_slice = &buffer.as_ref()[header.buffer_length()..];
+        let payload = match header.tag {
+            Tag::Sum => PayloadOwned::Sum(
+                SumOwned::from_bytes(&payload_slice).context("invalid sum payload")?,
+            ),
+            Tag::Update => PayloadOwned::Update(
+                UpdateOwned::from_bytes(&payload_slice).context("invalid update payload")?,
+            ),
+            Tag::Sum2 => PayloadOwned::Sum2(
+                Sum2Owned::from_bytes(&payload_slice).context("invalid sum2 payload")?,
+            ),
+        };
+        Ok(Self { header, payload })
+    }
+}
+
+/// A seal to sign and encrypt messages
+pub struct MessageSeal<'a, 'b> {
```

Comment:
i would suggest either `MessageSeal`/`MessageOpen` or `MessageSealer`/`MessageOpener` but not a mixture of both.

Reply:
Ah cool I didn't know that "sealer" as a noun existed. It sounds a bit odd to me though and seal seems more common: https://en.wikipedia.org/wiki/Seal_(emblem)

I'd be down to change to "sealer" though. Do you validate @finiteprods ?

Reply:
the `recipient_pk` of the coordinator is only valid for the current round, hence the participant will hold this information not much longer after the next round starts. in contrast, the `sender_sk` of the participant must be valid for all rounds.

Reply:
Yeah as @janpetschexain said this id just to express the fact that these two fields might have different lifetimes.